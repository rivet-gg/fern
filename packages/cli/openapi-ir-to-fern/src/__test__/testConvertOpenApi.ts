import { AbsoluteFilePath, join, RelativeFilePath } from "@fern-api/fs-utils";
import { CONSOLE_LOGGER } from "@fern-api/logger";
import { parse } from "@fern-api/openapi-parser";
import { createMockTaskContext } from "@fern-api/task-context";
import path from "path";
import { convert } from "../convert";

const FIXTURES_PATH = join(AbsoluteFilePath.of(__dirname), RelativeFilePath.of("fixtures"));

// eslint-disable-next-line jest/no-export
export function testConvertOpenAPI(fixtureName: string, filename: string, asyncApiFilename?: string): void {
    // eslint-disable-next-line jest/valid-title
    describe(fixtureName, () => {
        it("simple", async () => {
            const openApiPath = path.join(FIXTURES_PATH, fixtureName, filename);
            const mockTaskContext = createMockTaskContext({ logger: CONSOLE_LOGGER });

            const absolutePathToAsyncAPI =
                asyncApiFilename != null
                    ? join(FIXTURES_PATH, RelativeFilePath.of(fixtureName), RelativeFilePath.of(asyncApiFilename))
                    : undefined;

            const specs = [];
            specs.push({
                absoluteFilepath: AbsoluteFilePath.of(openApiPath),
                absoluteFilepathToOverrides: undefined
            });
            if (absolutePathToAsyncAPI != null) {
                specs.push({
                    absoluteFilepath: absolutePathToAsyncAPI,
                    absoluteFilepathToOverrides: undefined
                });
            }

            const openApiIr = await parse({
                specs,
                taskContext: createMockTaskContext({ logger: CONSOLE_LOGGER })
            });
            const fernDefinition = convert({
                ir: openApiIr,
                taskContext: mockTaskContext,
                enableUniqueErrorsPerEndpoint: false,
                detectGlobalHeaders: true
            });
            expect(fernDefinition).toMatchSnapshot();
        });

        it("docs", async () => {
            const openApiPath = path.join(FIXTURES_PATH, fixtureName, filename);
            const mockTaskContext = createMockTaskContext({ logger: CONSOLE_LOGGER });

            const absolutePathToAsyncAPI =
                asyncApiFilename != null
                    ? join(FIXTURES_PATH, RelativeFilePath.of(fixtureName), RelativeFilePath.of(asyncApiFilename))
                    : undefined;

            const specs = [];
            specs.push({
                absoluteFilepath: AbsoluteFilePath.of(openApiPath),
                absoluteFilepathToOverrides: undefined
            });
            if (absolutePathToAsyncAPI != null) {
                specs.push({
                    absoluteFilepath: absolutePathToAsyncAPI,
                    absoluteFilepathToOverrides: undefined
                });
            }

            const openApiIr = await parse({
                specs,
                taskContext: createMockTaskContext({ logger: CONSOLE_LOGGER })
            });
            const fernDefinition = convert({
                ir: openApiIr,
                taskContext: mockTaskContext,
                enableUniqueErrorsPerEndpoint: true,
                detectGlobalHeaders: false
            });
            expect(fernDefinition).toMatchSnapshot();
        });
    });
}
