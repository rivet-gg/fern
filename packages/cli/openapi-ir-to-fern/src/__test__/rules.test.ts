import { testConvertOpenAPI } from "./testConvertOpenApi";

describe("rules", () => {
    testConvertOpenAPI("rules", "openapi.yml");
});
