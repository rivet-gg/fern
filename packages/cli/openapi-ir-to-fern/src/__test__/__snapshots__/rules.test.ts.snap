// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules rules docs 1`] = `
{
  "definitionFiles": {},
  "packageMarkerFile": {
    "service": {
      "auth": false,
      "base-path": "",
      "endpoints": {
        "defaultAndValidationRules": {
          "auth": false,
          "display-name": "Default and validation rules",
          "docs": undefined,
          "examples": [
            {
              "response": {
                "body": {
                  "age": 1,
                  "id": 1,
                  "name": "name",
                  "value": 1.1,
                },
              },
            },
          ],
          "method": "GET",
          "pagination": undefined,
          "path": "/example",
          "response": {
            "docs": "Successful response",
            "type": "GetExampleResponse",
          },
        },
      },
    },
    "types": {
      "GetExampleResponse": {
        "discriminated": false,
        "docs": undefined,
        "union": [
          "Schema",
          "GetExampleResponseMessage",
        ],
      },
      "GetExampleResponseMessage": {
        "docs": undefined,
        "properties": {
          "message": "optional<string>",
        },
      },
      "Schema": {
        "docs": undefined,
        "properties": {
          "age": {
            "default": 50,
            "docs": "The age of the type.",
            "type": "optional<integer>",
            "validation": {
              "exclusiveMax": false,
              "exclusiveMin": true,
              "max": undefined,
              "min": undefined,
              "multipleOf": 10,
            },
          },
          "id": {
            "default": 50,
            "docs": "The unique ID of the type.",
            "type": "optional<integer>",
            "validation": {
              "exclusiveMax": false,
              "exclusiveMin": true,
              "max": 100,
              "min": 10,
              "multipleOf": 10,
            },
          },
          "name": {
            "default": "type",
            "docs": "The unique name of the type.",
            "type": "optional<string>",
            "validation": {
              "format": undefined,
              "maxLength": 10,
              "minLength": 3,
              "pattern": "^[a-zA-Z0-9]*$",
            },
          },
          "value": {
            "default": 1.1,
            "docs": "The value of the type.",
            "type": "optional<double>",
            "validation": {
              "exclusiveMax": undefined,
              "exclusiveMin": undefined,
              "max": 2.2,
              "min": 1.1,
              "multipleOf": 1.1,
            },
          },
        },
      },
      "String": {
        "type": "string",
        "validation": {
          "format": undefined,
          "maxLength": 10,
          "minLength": 3,
          "pattern": "^[a-zA-Z0-9]*$",
        },
      },
    },
  },
  "rootApiFile": {
    "display-name": "Sample API",
    "error-discrimination": {
      "strategy": "status-code",
    },
    "name": "api",
  },
}
`;

exports[`rules rules simple 1`] = `
{
  "definitionFiles": {},
  "packageMarkerFile": {
    "service": {
      "auth": false,
      "base-path": "",
      "endpoints": {
        "defaultAndValidationRules": {
          "auth": false,
          "display-name": "Default and validation rules",
          "docs": undefined,
          "examples": [
            {
              "response": {
                "body": {
                  "age": 1,
                  "id": 1,
                  "name": "name",
                  "value": 1.1,
                },
              },
            },
          ],
          "method": "GET",
          "pagination": undefined,
          "path": "/example",
          "response": {
            "docs": "Successful response",
            "type": "GetExampleResponse",
          },
        },
      },
    },
    "types": {
      "GetExampleResponse": {
        "discriminated": false,
        "docs": undefined,
        "union": [
          "Schema",
          "GetExampleResponseMessage",
        ],
      },
      "GetExampleResponseMessage": {
        "docs": undefined,
        "properties": {
          "message": "optional<string>",
        },
      },
      "Schema": {
        "docs": undefined,
        "properties": {
          "age": {
            "default": 50,
            "docs": "The age of the type.",
            "type": "optional<integer>",
            "validation": {
              "exclusiveMax": false,
              "exclusiveMin": true,
              "max": undefined,
              "min": undefined,
              "multipleOf": 10,
            },
          },
          "id": {
            "default": 50,
            "docs": "The unique ID of the type.",
            "type": "optional<integer>",
            "validation": {
              "exclusiveMax": false,
              "exclusiveMin": true,
              "max": 100,
              "min": 10,
              "multipleOf": 10,
            },
          },
          "name": {
            "default": "type",
            "docs": "The unique name of the type.",
            "type": "optional<string>",
            "validation": {
              "format": undefined,
              "maxLength": 10,
              "minLength": 3,
              "pattern": "^[a-zA-Z0-9]*$",
            },
          },
          "value": {
            "default": 1.1,
            "docs": "The value of the type.",
            "type": "optional<double>",
            "validation": {
              "exclusiveMax": undefined,
              "exclusiveMin": undefined,
              "max": 2.2,
              "min": 1.1,
              "multipleOf": 1.1,
            },
          },
        },
      },
      "String": {
        "type": "string",
        "validation": {
          "format": undefined,
          "maxLength": 10,
          "minLength": 3,
          "pattern": "^[a-zA-Z0-9]*$",
        },
      },
    },
  },
  "rootApiFile": {
    "display-name": "Sample API",
    "error-discrimination": {
      "strategy": "status-code",
    },
    "name": "api",
  },
}
`;
